network/iptables.go

```diff
@@ -16,145 +16,273 @@
-			{"nat", "POSTROUTING", []string{"-s", n, "-d", n, "-j", "RETURN"}},
+			{"nat", "POSTROUTING", []string{"-s", c.Network, "-d", c.Network, "-j", "RETURN"}, ""},
 			// NAT if it's not multicast traffic
-			{"nat", "POSTROUTING", []string{"-s", n, "!", "-d", "224.0.0.0/4", "-j", "MASQUERADE"}},
+			{"nat", "POSTROUTING", []string{"-s", c.Network, "!", "-d", "224.0.0.0/4", "-j", "MASQUERADE", "--random-fully"}, ""},
 			// Prevent performing Masquerade on external traffic which arrives from a Node that owns the container/pod IP address
-			{"nat", "POSTROUTING", []string{"!", "-s", n, "-d", sn, "-j", "RETURN"}},
+			{"nat", "POSTROUTING", []string{"!", "-s", c.Network, "-d", c.Lease, "-j", "RETURN"}, ""},
 			// Masquerade anything headed towards flannel from the host
-			{"nat", "POSTROUTING", []string{"!", "-s", n, "-d", n, "-j", "MASQUERADE"}},
-		}
+			{"nat", "POSTROUTING", []string{"!", "-s", c.Network, "-d", c.Network, "-j", "MASQUERADE", "--random-fully"}, ""},
+		}...)
 	}
-}
 
-func ForwardRules(flannelNetwork string) []IPTablesRule {
-	return []IPTablesRule{
+	rules = append(rules, []IPTablesRule{
+		// This rule makes sure we don't NAT traffic within overlay network (e.g. coming out of docker0)
+		{"nat", "POSTROUTING", []string{"-s", c.Network, "-d", c.Network, "-j", "RETURN"}, ""},
+		// NAT if it's not multicast traffic
+		{"nat", "POSTROUTING", []string{"-s", c.Network, "!", "-d", "224.0.0.0/4", "-j", "MASQUERADE"}, ""},
+		// Prevent performing Masquerade on external traffic which arrives from a Node that owns the container/pod IP address
+		{"nat", "POSTROUTING", []string{"!", "-s", c.Network, "-d", c.Lease, "-j", "RETURN"}, ""},
+		// Masquerade anything headed towards flannel from the host
+		{"nat", "POSTROUTING", []string{"!", "-s", c.Network, "-d", c.Network, "-j", "MASQUERADE"}, ""},
+	}...)
+
+	// legacy forwarding rules
+	rules = append(rules, []IPTablesRule{
 		// These rules allow traffic to be forwarded if it is to or from the flannel network range.
-		{"filter", "FORWARD", []string{"-s", flannelNetwork, "-j", "ACCEPT"}},
-		{"filter", "FORWARD", []string{"-d", flannelNetwork, "-j", "ACCEPT"}},
-	}
+		{"filter", "FORWARD", []string{"-s", c.Network, "-j", "ACCEPT"}, ""},
+		{"filter", "FORWARD", []string{"-d", c.Network, "-j", "ACCEPT"}, ""},
+	}...)
+
+	return rules
 }
 
-func ipTablesRulesExist(ipt IPTables, rules []IPTablesRule) (bool, error) {
-	for _, rule := range rules {
-		exists, err := ipt.Exists(rule.table, rule.chain, rule.rulespec...)
+func (c Config) cleanupRules(ipt IPTables) {
+	for _, chain := range chains {
+		glog.Info("Clearing chain: table: ", chain.table, " chain: ", chain.name)
+
+		err := ipt.ClearChain(chain.table, chain.name)
 		if err != nil {
-			// this shouldn't ever happen
-			return false, fmt.Errorf("failed to check rule existence: %v", err)
-		}
-		if !exists {
-			return false, nil
+			glog.Info("Clear chain ", chain, " failed: ", err)
```

Comment:
nit: is this supposed to be just informational?

Reply:
I'll up it to warning, just worried someone might get confused because if it's a temporary failure it should reconcile later.